### PR TITLE
BUG: use SOS filters in decimate for numerical stability

### LINKED ISSUE #13529: signal.decimate doesn't use sosfilters and sosfiltfilt
Since the default recommended filters for scipy are now SOS, it should make sense to use them for decimate (and other internal uses)

This is the current state of scipy.signal.decimate
```
        if zero_phase:
            y = filtfilt(b, a, x, axis=axis)
        else:
            y = lfilter(b, a, x, axis=axis)
```

Sorry if this isn't the right class of Issue, new to doing this.

### LINKED ISSUE #15072: BUG: signal.decimate returns NaN with large float32 arrays
### Describe your issue.

When using decimate with large float32 arrays often arrays will all NaN values are returned.

### Reproducing Code Example

```python
import numpy as np
from scipy.signal import decimate

ar = np.random.random((10_000, 2))

out1 = decimate(ar, 10, axis=0)  # this works fine 

out2 = decimate(ar.astype(np.float32), 10, axis=0)  # returns all NaN

# It also works fine if the arrays smaller

out3 = decimate(ar.astype(np.float32)[::100], 10, axis=0)  # No NaNs in array
```

### Error message

```shell
None
```

### SciPy/NumPy/Python version information

scipy: 1.6.2, Numpy: 1.18.5, Python: 3.8.5

### LINKED ISSUE #15393: BUG: signal.decimate returns unexpected values with float32 arrays
### Describe your issue.

This may be a duplicate of gh-15072, but 1) I don't have a big array (50 values), and 2) this method did not return any NaN's, just unexpected large (negative or positive) values.

I finally caught the issue when I realized I had switched from scipy==1.4.1 to scipy==1.7.3 (before, it was working)

I'm attaching two plots made with the same code example, but each with a different scipy version. In the 1.4.1 one, the decimate result is correct and independent of the type of array, and in the 1.7.3, it's wrong for the float32 array:

scipy1
scipy2

### Reproducing Code Example

```python
import scipy.signal as ssi
import matplotlib.pyplot as plt
import numpy as np

np.random.seed(1)
n = 50
p = np.random.rand(n)
q = 20
p_d1 = ssi.decimate(np.asarray(p, dtype='float32'), q)
p_d2 = ssi.decimate(np.asarray(p, dtype='float64'), q)
print('max p_d1 (float32): ', max(p_d1))  # scipy==1.4.1: 0.526 / scipy==1.7.3: 5279  <- wrong
print('max p_d2 (float64): ', max(p_d2))  # scipy==1.4.1: 0.526 / scipy==1.7.3: 0.526

# Optionally plot
t = list(range(n))
t_d = [t[i * q] for i in range(len(p_d1))]
plt.plot(t, p, marker='o')
plt.plot(t_d, p_d1, marker='s')
plt.plot(t_d, p_d2, marker='x')
plt.legend(['float32', 'float32 > decimated', 'float64 > decimated'], loc='upper right')
plt.title('scipy==1.4.1')  # scipy==1.4.1 / scipy==1.7.3 (latest to date)
plt.show()
```

### Error message

```shell
None!!
```

### SciPy/NumPy/Python version information

No issue with scipy==1.4.1; Issue with scipy==1.7.3

### PR DESCRIPTION
**Pull request by roryyorke**
#### Reference issue
Fixes gh-13529, fixes gh-15072, and fixes gh-15393.

#### What does this implement/fix?
Use `sosfilt` and `sosfiltfilt` instead of, respectively, `lfilter` and `filtfilt`, for better numerical stability.

#### Additional information
This should produce the same output as previously, except for 32-bit floats where the previous code was numerically unstable.  There may also be differences in output if the caller passed in an LTI object that was marginally stable and is now becomes stable.

I added a regression test for the 32-bit float case, based on gh-15393.

I did *not* update `TestDecimate._test_phaseshift` in `test_signaltools.py`; that still tests the behaviour of `zero_phase`, but not with the actual filter synthesis now used in `decimate`.  It looks to me like it was already out-of-date w.r.t. the FIR case, or at least the code for that is different.

### CODE REVIEW

File: scipy/signal/_signaltools.py
Comment (ilayn) [inline scipy/signal/_signaltools.py:4461]: you can use `if x.dtype is not np.inexact:` test here which also covers `np.float16` case.

Comment (roryyorke) [reply scipy/signal/_signaltools.py:4461]: OK - I didn't actually change this behaviour, just moved it earlier (needed for different FIR & IIR handling).  It seems like a reasonable change.

Comment (larsoner) [reply scipy/signal/_signaltools.py:4461]: @roryyorke want to try this change and add a `float16` case in the tests somewhere? Then this LGTM +1 for merge!

**Comment by roryyorke** [comment]
Neither `signal.lfilter` nor `signal.sosfilt` currently support `np.float16`:

```
scipy/signal/_signaltools.py:2069: in lfilter
    return _sigtools._linear_filter(b, a, x, axis)
E   NotImplementedError: input type 'float16' not supported
```

```
scipy/signal/_signaltools.py:4227: in sosfilt
    raise NotImplementedError("input type '%s' not supported" % dtype)
E   NotImplementedError: input type 'float16' not supported
```

`sosfiltfilt` calls `sosfilt_zi`, which uses `np.linalg.solve`, and gives:
```
E   TypeError: array type float16 is unsupported in linalg
```

There may be other things that don't work; I stopped looking.

It looks like adding "native" float16 support to the IIR-filtering capabilities of `scipy.signal` will be a non-trivial project.  There's https://github.com/numpy/numpy/issues/11690 on the numpy side, which suggests upcasting (I presume to float32); this would probably be OK for scipy `sosfilt_zi` and any other "support" calcs done for filtering.  

For actual filtering (the "for-loop" running over the input array), I presume float16 support is needed at the C level, or one must again upcast to float32.

**Comment by ilayn** [comment]
Maybe my comment was misleading. By the change I proposed we would be upcasting float 16 to float 64. Indeed we don't want to work with float16 anyways

**Comment by roryyorke** [comment]
Ah, I see.  Upcast to float64 or float32?  I'd naively expect 32 (the "next-best" option), but I've never worked with 16-bit floats.

I added this:

```python
    if not issubclass(result_type.type, np.inexact):
        result_type = np.float64
```

I presume this would change to

```
    if not issubclass(result_type.type, np.inexact) or result_type.type == np.float16:
        result_type = np.float64
```

or, perhaps,

```
    if not issubclass(result_type.type, [np.float32, np.float64, np.complex128]):
        result_type = np.float64
```

**Comment by ilayn** [comment]
It's even better, instead of 

```python
    if not issubclass(result_type.type, np.inexact):
        result_type = np.float64
```

this is sufficient 

```python
   if x.dtype is not np.inexact:
        result_type = np.float64
```

and since np.float16 is not based on floating point type and it is a franken-type of unsigned 16 or something like that, it will also catch np.float16 and upcast it to float64.

**Comment by roryyorke** [comment]
I don't understand the `x.dtype is not np.inexact` condition; in particular, I don't think it's possible to override `is` in Python's data model; this is why I used `issubclass` before I was aware of `float16`'s special behaviour.

```
In [12]: x = np.zeros(1,dtype=np.float64)

In [13]: x.dtype is not np.inexact
Out[13]: True

In [14]: x = np.zeros(1,dtype=np.int8)

In [15]: x.dtype is not np.inexact
Out[15]: True

In [16]: np.__version__
Out[16]: '1.22.3'
```

What am I missing?

Also, `np.float128` and `np.complex256` don't work with `filtfilt` and `sosfiltfilt` either: should we special-case them too, or just allow the errors to be raised?

**Comment by roryyorke** [comment]
I've committed what I think is the right thing for upcasting np.float16 arguments to np.float64.

I also used `numpy.issubdtype` instead of `issubclass`; from a brief scan of scipy code, `issubdtype` is more idiomatic.

**Comment by roryyorke** [comment]
I couldn't reproduce the flake8 (I assume) E127 lint error here.

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge

@ilayn feel free to merge if you're happy

**Review by tylerjereddy — COMMENTED** [review]
This has a core developer approval. There is a little bit of debate about refinements to the type handling, but as far as I can tell this is still fixing a bug and has new regression tests with no old test modified, which is usually a good sign.

This hasn't been flushed through CI in a while--I rebased locally and it did pass the full suite, but I've pushed the rebase in to make sure CI is happy as well.

**Comment by tylerjereddy** [comment]
It may be good to double check that all 3 issues were indeed closed by this.

**Comment by roryyorke** [comment]
#13529 is "decimate doesn't use sosfilt&sosfiltfilt"; it's fixed :)

#15072 (large float32 arrays, high q gives NaNs):

```python
In [1]: import scipy

In [2]: scipy.__version__
Out[2]: '1.10.0.dev0+2280.276f9e0'

In [3]: from scipy.signal import decimate

In [4]: import numpy as np

In [5]: ar = np.random.random((10_000, 2))

In [6]: np.any(np.isnan(decimate(ar.astype(np.float32), 10, axis=0)))
Out[6]: False
```

#15393 (particular example gave wrong max with float32), continuing session above:

```python
In [9]: import scipy.signal as ssi
   ...: import matplotlib.pyplot as plt
   ...: import numpy as np
   ...:
   ...: np.random.seed(1)
   ...: n = 50
   ...: p = np.random.rand(n)
   ...: q = 20
   ...: p_d1 = ssi.decimate(np.asarray(p, dtype='float32'), q)
   ...: p_d2 = ssi.decimate(np.asarray(p, dtype='float64'), q)
   ...: print('max p_d1 (float32): ', max(p_d1))  # scipy==1.4.1: 0.526 / scipy==1.7.3: 5279  <- wrong
   ...: print('max p_d2 (float64): ', max(p_d2))  # scipy==1.4.1: 0.526 / scipy==1.7.3: 0.526
max p_d1 (float32):  0.5261423
max p_d2 (float64):  0.5261338016034416
```

**Comment by larsoner** [comment]
Thanks for double checking @roryyorke !

**Pull request merged**
